ptvsd/runner.py

```diff
@@ -0,0 +1,359 @@
+                'nodebug': True
+            },
+        )
+
+    # closing the adapter
+
+    def close(self):
+        """Stop the message processor and release its resources."""
+        if self._closed:
+            return
+        self._closed = True
+
+        self._notify_closing()
+        # Close the editor-side socket.
+        self._stop_vsc_message_loop()
+
+    def _stop_vsc_message_loop(self):
+        self.set_exit()
+        if self._socket:
+            try:
+                self._socket.shutdown(socket.SHUT_RDWR)
+                self._socket.close()
+            except:
+                pass
+
+    def _wait_for_server_thread(self):
+        if self.server_thread is None:
+            return
+        if not self.server_thread.is_alive():
+            return
+        self.server_thread.join(WAIT_FOR_THREAD_FINISH_TIMEOUT)
+
+    def handle_stopped(self, exitcode):
+        """Finalize the protocol connection."""
+        if self._exited:
+            return
+        self._exited = True
+
+        # Notify the editor that the "debuggee" (e.g. script, app) exited.
+        self.send_event('exited', exitCode=exitcode)
+
+        # Notify the editor that the debugger has stopped.
+        self.send_event('terminated')
+
+        # The editor will send a "disconnect" request at this point.
+        self._wait_for_disconnect()
+
+    def _wait_for_disconnect(self, timeout=None):
+        if timeout is None:
+            timeout = WAIT_FOR_DISCONNECT_REQUEST_TIMEOUT
+
+        if not self.disconnect_request_event.wait(timeout):
+            warnings.warn('timed out waiting for disconnect request')
+        if self.disconnect_request is not None:
+            self.send_response(self.disconnect_request)
+            self.disconnect_request = None
+
+    def _handle_disconnect(self, request):
+        self.disconnect_request = request
+        self.disconnect_request_event.set()
+        self._notify_disconnecting(not self._closed)
+        if not self._closed:
+            self.close()
+
+    # VSC protocol handlers
+
+    def on_initialize(self, request, args):
+        self.send_response(
+            request,
+            supportsExceptionInfoRequest=True,
+            supportsConfigurationDoneRequest=True,
+            supportsConditionalBreakpoints=True,
+            supportsSetVariable=True,
+            supportsExceptionOptions=True,
+            supportsEvaluateForHovers=True,
+            supportsValueFormattingOptions=True,
+            supportsSetExpression=True,
+            supportsModulesRequest=True,
+            exceptionBreakpointFilters=[
+                {
+                    'filter': 'raised',
+                    'label': 'Raised Exceptions',
+                    'default': False
+                },
+                {
+                    'filter': 'uncaught',
+                    'label': 'Uncaught Exceptions',
+                    'default': True
+                },
+            ],
```

Comment:
Yet another copy of this. I created a work item to create a single location for these values. This is now duplicated in too many places.
[https://github.com/Microsoft/ptvsd/issues/342](https://github.com/Microsoft/ptvsd/issues/342)

Reply:
This will not be used by VSC when running, but I see your point.
Thanks for creating that ticket.

Reply:
You'll notice that I'm returning `success` for almost all requests, that's because VSC doesn't really use them, except of course when `success=false`